### ISSUE TITLE
Make additional zmq constants and procs scoped public

### ISSUE THREAD

**Issue by hokiegeek2** [chapel-lang/chapel#15781]
# Summary of Problem
Just a few minor changes to the ZMQ package are needed. There are a few  ZMQ constants and procs I need to make accessible from the public Chapel API in analogy to what's been done previously with ZMQ_PUB and zmq_send: 

`/*
    The publisher socket type for a publish-subscribe messaging pattern.
   */
  const PUB  = ZMQ_PUB;`

`proc send(data: ?T, flags: int = 0) throws where isNumericType(T) {
      on classRef.home {
        var copy = data;
        while (-1 == zmq_send(classRef.socket, c_ptrTo(copy):c_void_ptr,
                              numBytes(T):size_t,
                              (ZMQ_DONTWAIT | flags):c_int)) {
          if errno == EAGAIN then
            chpl_task_yield();
          else {
            try throw_socket_error(errno, "send");
          }
        }
      }
    }`

The changes are needed to enable basic authentication in the [Arkouda](https://github.com/mhmerrill/arkouda) project.

I discussed this with @bradcray @ronawho and will be happy to fork chapel and make the changes myself.

**Comment by bradcray** [comment]
@nspark / @lydia-duncan: Do either of you remember why so many `extern const`s are marked `private` in ZMQ.chpl, but not used?  (an unused `private extern const` seems pointless to me, but maybe I'm missing something?).  As an example, most of the following seem unused:

https://github.com/chapel-lang/chapel/blob/bcee2d37f227c0cac0fb4ef3ae4807d96736b791/modules/packages/ZMQ.chpl#L365-L414

**Comment by lydia-duncan** [comment]
I think the intention was that they would eventually be useful but that the focus was on ones that were immediately relevant.  I don't remember, was the ZMQ.chpl file created using c2chapel?  That would make it much easier to generate unused ones, meaning that explicitly removing them would be more effort than just including them, right?

**Comment by lydia-duncan** [comment]
In any case, I would be in favor of expanding the functionality and definitely down to review such changes.  It's possible that building off #15283 would be useful, though I'm not sure right now if the contributor intends to finish that work (I've pinged them again and if they don't respond I will close the PR)

**Comment by hokiegeek2** [comment]
@lydia-duncan argh, sorry, I should have searched the open issues first to see if there were similar requests, thanks for bringing that one to my attention. These minor changes are in my critical path, so I am happy to work this myself.

**Comment by lydia-duncan** [comment]
>  In ZMQ it looks like they're used to send into a setSockOpt type of call, but are we taking the approach of making distinct calls per option rather than a very generic call like that in Chapel? (this is reminding me of conversations from awhile back, but I can't quite get all the context back...).

Yeah, the intention is to expose to the user distinct calls per option, but we'd still need to use those constants under the covers to call to the extern proc, like we do for getLastEndpoint [here](https://github.com/chapel-lang/chapel/blob/master/modules/packages/ZMQ.chpl#L719)

**Comment by hokiegeek2** [comment]
@bradcray There's actually more to this than the issue of public vs private constants. Basically, authentication is not implemented. I am going to rename this issue and detail the full scope of what has to be done.

**Comment by hokiegeek2** [comment]
@bradcray @ronawho I was able to add the constants and the setter methods for PLAIN ZMQ authentication in my [chapel fork](https://github.com/hokiegeek2/chapel/commits/15781_make_additional_zmq_constants_and_procs_scoped_public). The next thing I need to do is figure out the syntax for starting up the auth actor.

**Comment by bradcray** [comment]
Sounds good @hokiegeek2, let us know if we can help by answering questions / providing pointers.
